### ISSUE TITLE
`do -i` Printing stderr from External Commands

### ISSUE THREAD

**Issue by uvb1** [nushell/nushell#12451]
### Question

Before v0.92.0

`do -i { diff nonexistentfile1 nonexistentfile2 }` resulted in nothing output, as expected.

After v0.92.0

`do -i { diff nonexistentfile1 nonexistentfile2 }`  results in the following being output: `diff: nofile1: No such file or directory`

Is this change in behavior expected?

**Comment by devyn** [comment]
Yes, this is expected. `-i` has nothing to do with stderr, it's all about Nushell errors (the ones you can catch with `try`)

There was a flaw before where the stream printer would only print stdout from a captured stream even if stderr was available. This has been fixed so now you see stderr too. If you pipe to another command you will still only see stdout, unless it explicitly uses stderr too or you use a redirection like `o+e>|` to merge them together.

**Comment by fdncred** [comment]
Also, it's typical that `do -i` gets piped to `complete` so one can inspect errors in a structured way.
```nushell
❯ do -i { diff no file } | complete
╭───────────┬───────────────────────────────────────╮
│ stdout    │                                       │
│ stderr    │ diff: no: No such file or directory   │
│           │ diff: file: No such file or directory │
│           │                                       │
│ exit_code │ 2                                     │
╰───────────┴───────────────────────────────────────╯
```

Having said that, I think we need to look into this further because it is working differently now. Good catch @uvb1!

**Comment by fdncred** [comment]
@IanManske What are your thoughts here? I think `do -i` stopped working the way it used to after https://github.com/nushell/nushell/pull/11934

**Comment by IanManske** [comment]
The `help` text for `do` isn't completely clear about what `-i` is supposed to do. Looking through the code, it looks like `-i` basically means `-s` and `-p` combined. Based on one of the `help do` examples, it looks like `-p` is meant to suppress non-zero exit codes. So, I think printing stderr should be expected. Command stderr output can always be redirected or ignored anyways:
```nushell
use std null-device
# ignore all stderr
do -i { diff nonexistentfile1 nonexistentfile2 } e> (null-device)
# or just for one command
do -i { diff nonexistentfile1 nonexistentfile2 e> (null-device) }
```
In the future, it might be worth to make `null-device` a command, or at least available in the global scope / automatically imported.

**Comment by fdncred** [comment]
@IanManske We should talk about this in our meeting on Wednesday because the expectation with `do -i` as far as i remember has always been to eat all errors, thus the name `--ignore-errors`. It's not `--ignore-some-errors-and-print-others`, lol. I agree with you that `-i` is the combination of `-s` and `-p`. My preference would be to maintain the ability to eat all errors with `-i`. Others may disagree.

**Comment by IanManske** [comment]
Some external commands use stderr for other purposes like logging (e.g., `cargo b`). So, I don't think we should suppress stderr output by treating it as an error. But can I see how suppressing all stderr output is a common use case, so I think we should make it more ergonomic for users to do so.

**Comment by AntoniosBarotsis** [comment]
An interesting side effect of this to consider, I've been using [panache-git](https://github.com/nushell/nu_scripts/blob/main/modules/prompt/panache-git.nu) and [this](https://github.com/nushell/nu_scripts/blob/c30efc727b99ee75ea0140375cf4651268fbe36a/modules/prompt/panache-git.nu#L46) part now makes it so `fatal: not a git repository (or any of the parent directories): .git` is printed every time in my shell. I'm just piping it to null-device for now as Ian suggested

**Comment by fdncred** [comment]
After discussing this in our team meeting yesterday, our decision was to try and simplify `do` and get rid of these parameters.
- `--ignore-errors/-i`
- `--ignore-shell-errors/-s`
- `--ignore-program-errors/-p`
- `--capture-errors/-c`

And replace them with better syntax for handling redirection. There's no ETA yet but that is the desired goal. In the mean time, I'll land the patch for panache-git that fixes the problem we introduced.

We may want to restore the `do -i` "eat all errors" functionality temporarily too.

**Comment by WindSoilder** [comment]
Closing the issue because the question is answered.
